fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinLAsscomHelper.java

```diff
@@ -0,0 +1,219 @@
+import org.apache.doris.nereids.trees.expressions.NamedExpression;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.plans.GroupPlan;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+import org.apache.doris.nereids.util.ExpressionUtils;
+import org.apache.doris.nereids.util.Utils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Common function for JoinLAsscom
+ */
+public class JoinLAsscomHelper {
+    private final LogicalJoin topJoin;
+    private final LogicalJoin<GroupPlan, GroupPlan> bottomJoin;
+    private final Plan a;
+    private final Plan b;
+    private final Plan c;
+
+    private final Expression topJoinOnClause;
+    private final Expression bottomJoinOnClause;
+
+    private final List<SlotReference> aOutputSlots;
+    private final List<SlotReference> bOutputSlots;
+    private final List<SlotReference> cOutputSlots;
+
+    private final List<Expression> newBottomJoinOnCondition = Lists.newArrayList();
+    private final List<Expression> newTopJoinOnCondition = Lists.newArrayList();
+
+    /**
+     * Init plan and output.
+     */
+    public JoinLAsscomHelper(LogicalJoin topJoin, LogicalJoin<GroupPlan, GroupPlan> bottomJoin) {
+        this.topJoin = topJoin;
+        this.bottomJoin = bottomJoin;
+
+        a = bottomJoin.left();
+        b = bottomJoin.right();
+        c = (Plan) topJoin.right();
+
+        Preconditions.checkArgument(topJoin.getCondition().isPresent(), "topJoin onClause must be present.");
+        topJoinOnClause = (Expression) topJoin.getCondition().get();
+        Preconditions.checkArgument(bottomJoin.getCondition().isPresent(), "bottomJoin onClause must be present.");
+        bottomJoinOnClause = (Expression) bottomJoin.getCondition().get();
+
+        aOutputSlots = Utils.getOutputSlotReference(a);
+        bOutputSlots = Utils.getOutputSlotReference(b);
+        cOutputSlots = Utils.getOutputSlotReference(c);
+    }
+
+    public static JoinLAsscomHelper of(LogicalJoin topJoin, LogicalJoin<GroupPlan, GroupPlan> bottomJoin) {
+        return new JoinLAsscomHelper(topJoin, bottomJoin);
+    }
+
+    /**
+     * Get the onCondition of newTopJoin and newBottomJoin.
+     */
+    public boolean initJoinOnCondition() {
+        List<Expression> topJoinOnClauseConjuncts = ExpressionUtils.extractConjunction(topJoinOnClause);
+        for (Expression topJoinOnClauseConjunct : topJoinOnClauseConjuncts) {
+            // Ignore join with some OnClause like:
+            // Join C = B + A for above example.
+            if (ExpressionUtils.isIntersecting(
+                    topJoinOnClauseConjunct.collect(SlotReference.class::isInstance), aOutputSlots)
```

Comment:
we should do `topJoinOnClauseConjunct.collect(SlotReference.class::isInstance)` only once outside if statement